environment.go

```diff
@@ -378,6 +387,85 @@ func (c *envState) GetDB() *mongo.Database {
 	return c.client.Database(c.conf.DatabaseName)
 }
 
+// RegisterDBValueCacher adds a new value to the database cache with the given
+// name as the key. It also accepts a channel which it uses to receive updated
+// values in a goroutine. Once an updated value is received, the previous
+// value stored in the cache is replaced with the new one. If an error is
+// received on the channel, the value is removed from the cache entirely and
+// the goroutine exits. This function returns a struct{} channel used to
+// notify the caller that this value cacher is no longer listening on its
+// update channel (i.e. the goroutine exited) and a boolean indicating if the
+// value was successfully cached. False is returned if either the database
+// cache is disabled or a key with the given name already exists in the cache.
+//
+// Note: The database value cache is thread safe.
+func (c *envState) RegisterDBValueCacher(name string, val interface{}, updateChan chan interface{}) (chan struct{}, bool) {
```

Comment:
Rather than relying on a channel closing to tell the change stream when to stop, it seems like the only exit case in both goroutines is the environment's context cancelling (panics are not exit cases since there's a deferred recover and continue). Can we just have `changeStream` use the environment context and not accept the `closeChan`?

I would also pass `c.env.Context()` to both `RegisterDBValueCacher and the change stream goroutine to ensure they share the same context.

Reply:
> Can we separate the error channel and success update channel in the input parameters? It seems like it would be cleaner if you knew that receiving an update meant the value is updated and receiving an error means the cache updater cannot continue.

I like keeping a single "updates" channel (an error is a valid update), especially since it is an interface{} channel. Having so many channels will be even more confusing in my opinion.

> Rather than relying on a channel closing to tell the change stream when to stop, it seems like the only exit case in both goroutines is the environment's context cancelling (panics are not exit cases since there's a deferred recover and continue). Can we just have changeStream use the environment context and not accept the closeChan?
>
> I would also pass c.env.Context() to both `RegisterDBValueCacher and the change stream goroutine to ensure they share the same context.

I am not sure what you mean by "panics are not exit cases". If the thread exits due to a panic, the channel will get closed and notify the caller that the thread is no longer listening on the update channel (despite the env context not getting canceled).